src/source-dpdk.c

```diff
@@ -0,0 +1,494 @@
+}
+
+/* get number of seconds from the reset of TSC counter (typically from the machine start) */
+static uint64_t DPDKGetSeconds()
+{
+    return CyclesToSeconds(rte_get_tsc_cycles());
+}
+
+/**
+ * \brief Registration Function for RecieveDPDK.
+ * \todo Unit tests are needed for this module.
+ */
+void TmModuleReceiveDPDKRegister(void)
+{
+    tmm_modules[TMM_RECEIVEDPDK].name = "ReceiveDPDK";
+    tmm_modules[TMM_RECEIVEDPDK].ThreadInit = ReceiveDPDKThreadInit;
+    tmm_modules[TMM_RECEIVEDPDK].Func = NULL;
+    tmm_modules[TMM_RECEIVEDPDK].PktAcqLoop = ReceiveDPDKLoop;
+    tmm_modules[TMM_RECEIVEDPDK].PktAcqBreakLoop = NULL;
+    tmm_modules[TMM_RECEIVEDPDK].ThreadExitPrintStats = ReceiveDPDKThreadExitStats;
+    tmm_modules[TMM_RECEIVEDPDK].ThreadDeinit = ReceiveDPDKThreadDeinit;
+    tmm_modules[TMM_RECEIVEDPDK].cap_flags = SC_CAP_NET_RAW;
+    tmm_modules[TMM_RECEIVEDPDK].flags = TM_FLAG_RECEIVE_TM;
+}
+
+/**
+ * \brief Registration Function for DecodeDPDK.
+ * \todo Unit tests are needed for this module.
+ */
+void TmModuleDecodeDPDKRegister(void)
+{
+    tmm_modules[TMM_DECODEDPDK].name = "DecodeDPDK";
+    tmm_modules[TMM_DECODEDPDK].ThreadInit = DecodeDPDKThreadInit;
+    tmm_modules[TMM_DECODEDPDK].Func = DecodeDPDK;
+    tmm_modules[TMM_DECODEDPDK].ThreadExitPrintStats = NULL;
+    tmm_modules[TMM_DECODEDPDK].ThreadDeinit = DecodeDPDKThreadDeinit;
+    tmm_modules[TMM_DECODEDPDK].cap_flags = 0;
+    tmm_modules[TMM_DECODEDPDK].flags = TM_FLAG_DECODE_TM;
+}
+
+static inline void DPDKDumpCounters(DPDKThreadVars *ptv)
+{
+#ifdef PACKET_STATISTICS
+    // Prevents access to other fields of stats structure
+    if (unlikely(ptv->queue_id >= RTE_ETHDEV_QUEUE_STAT_CNTRS))
+        return;
+
+    struct rte_eth_stats eth_stats;
+    rte_eth_stats_get(ptv->port_id, &eth_stats);
+
+    uint64_t th_pkts = StatsGetLocalCounterValue(ptv->tv, ptv->capture_dpdk_packets);
+    uint64_t th_drops = StatsGetLocalCounterValue(ptv->tv, ptv->capture_dpdk_drops);
+
+    StatsAddUI64(ptv->tv, ptv->capture_dpdk_packets, eth_stats.q_ipackets[ptv->queue_id] - th_pkts);
+    StatsAddUI64(ptv->tv, ptv->capture_dpdk_drops, eth_stats.q_errors[ptv->queue_id] - th_drops);
+    SC_ATOMIC_ADD(ptv->livedev->pkts, eth_stats.q_ipackets[ptv->queue_id] - th_pkts);
+    SC_ATOMIC_ADD(ptv->livedev->drop, eth_stats.q_errors[ptv->queue_id] - th_drops);
+#endif
+}
+
+static void DPDKReleasePacket(Packet *p)
+{
+    int retval;
+    /* Need to be in copy mode and need to detect early release
+       where Ethernet header could not be set (and pseudo packet) */
+    if ((p->dpdk_v.copy_mode != DPDK_COPY_MODE_NONE) && !PKT_IS_PSEUDOPKT(p) &&
+            p->dpdk_v.copy_mode == DPDK_COPY_MODE_IPS && !PACKET_TEST_ACTION(p, ACTION_DROP)) {
+        retval =
+                rte_eth_tx_burst(p->dpdk_v.out_port_id, p->dpdk_v.out_queue_id, &p->dpdk_v.mbuf, 1);
+        if (unlikely(retval < 1)) {
+            rte_pktmbuf_free(p->dpdk_v.mbuf);
```

Comment:
either retry once more or update counter for each fail for transmit. rework

Reply:
there are lot easier ways to 

1. increment counter per worker
2. `But I added a second chance to transmit the packe` I assume this is the new change
3. `I additionally added a debug message that it was unable to transmit the packet on the given port and queue_id.` incorrect way.

note: just above you mentioned about latency and overhead now you are adding debug message will not increase latency (I in my testing it will). Please rework

Reply:
> I assume this is the new change

yes

> I additionally added a debug message that it was unable to transmit the packet on the given port and queue_id. incorrect way.

Debug messages are only displayed when Suricata is configured and compiled with the "Debug flag". Additionally, the log level must be set to DEBUG. Therefore in an ordinary use, no message will be displayed or logged..